lib/gollum-lib/file.rb

```diff
@@ -41,24 +44,32 @@ def path_compare(query, match_path, hyphened_tags, case_insensitive)
     # version - The String version ID to find.
     # try_on_disk - If true, try to return just a reference to a file
     #               that exists on the disk.
-    # global_match - If true, find a File matching path's filename, but not it's directory (so anywhere in the repo)
+    # global_match - If true, find a File matching path's filename, but not its directory (so anywhere in the repo)
     #
     # Returns a Gollum::File or nil if the file could not be found. Note
     # that if you specify try_on_disk=true, you may or may not get a file
     # for which on_disk? is actually true.
     def self.find(wiki, path, version, try_on_disk = false, global_match = false)
-      map = wiki.tree_map_for(version.to_s)
-
-      query_path = Pathname.new(::File.join(['/', wiki.page_file_dir, path].compact)).cleanpath.to_s
-      query_path.sub!(/^\/\//, '/') if Gem.win_platform? # On Windows, Pathname#cleanpath will leave double slashes at the start of a path intact, so sub them out.
+      query_path =  self.canonical_path(path, page_file_dir: wiki.page_file_dir)
+      path_segments = query_path.split('/')
+      filename = path_segments.last
+      dir = path_segments[0..-2].join('/')
 
-      begin
-        entry = map.detect do |entry|
-          path_match(query_path, entry, global_match, wiki.hyphened_tag_lookup, wiki.case_insensitive_tag_lookup)
+      if global_match && self.respond_to?(:global_find) # Only implemented for Gollum::Page
+        return self.global_find(wiki, version, query_path, try_on_disk) 
+      else
+        begin
+          root = wiki.commit_for(version.to_s)
```

Comment:
What kind of data structure is `version` that it needs the `to_s` call?

Reply:
It's a `Gollum::Git::Commit`. We could use `version.id` (which is an alias of `to_s`) instead?

Reply:
Actually, according to the inline docs it should be a `String`. We may be assuming that it *can* also be a Commit, in which case `to_s` is an option that works in either case. What do you think is better?

Reply:
I've opened [this issue](https://github.com/gollum/gollum/issues/1972) to resolve the ambiguity.